File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetMenu.java
package com.sup.dev.android.views.widgets;

import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.StringRes;
import android.support.v7.widget.LinearLayoutManager;

import com.sup.dev.android.tools.ToolsResources;
import com.sup.dev.android.views.adapters.recycler_view.RecyclerCardAdapter;
import com.sup.dev.android.views.cards.CardDivider;
import com.sup.dev.android.views.cards.CardDividerTitle;
import com.sup.dev.android.views.cards.CardMenu;
import com.sup.dev.java.classes.callbacks.simple.Callback1;
import com.sup.dev.java.classes.callbacks.simple.Callback2;
import com.sup.dev.java.classes.providers.Provider;

public class WidgetMenu extends WidgetRecycler {

    private final RecyclerCardAdapter adapter;
    private Callback2<WidgetMenu, String> onGlobalSelected;

    private int prefCount = 0;

    public WidgetMenu() {
        vRecycler.setLayoutManager(new LinearLayoutManager(view.getContext()));
        adapter = new RecyclerCardAdapter();
        setAdapter(adapter);
    }

    @Override
    public void onShow() {
        super.onShow();
        finishItemBuilding();
    }

    public void clear(){
        finishItemBuilding();
        adapter.clear();
        prefCount = 0;
    }

    private void add(Item item) {

        item.card = new CardMenu();
        item.card.setText(item.text);
        item.card.setIcon(item.icon);
        item.card.setBackground(item.bg);
        item.card.setOnClick((v, x, y) -> {
            hide();
            if (item.onClick != null) item.onClick.callback(this);
            if (onGlobalSelected != null) onGlobalSelected.callback(this, item.text);
        });


        if (item.preferred) {
            if (prefCount == 0) adapter.add(0, new CardDivider());
            adapter.add(prefCount, item.card);
            prefCount++;
        } else {
            adapter.add(item.card);
        }

    }

    public WidgetMenu group(@StringRes int title) {
        return group(ToolsResources.getString(title));
    }

    public WidgetMenu group(@StringRes int title, boolean divider) {
        return group(ToolsResources.getString(title), divider);
    }

    public WidgetMenu group(String title) {
        return group(title, true);
    }

    public WidgetMenu group(String title, boolean divider) {
        finishItemBuilding();
        adapter.add(new CardDividerTitle().setText(title).setDivider(divider));
        return this;
    }

    public void setOnGlobalSelected(Callback2<WidgetMenu, String> onGlobalSelected) {
        this.onGlobalSelected = onGlobalSelected;
    }

    @Override
    public WidgetMenu setTitle(@StringRes int title) {
        return super.setTitle(title);
    }

    @Override
    public WidgetMenu setTitle(String title) {
        return super.setTitle(title);
    }

    @Override
    public WidgetMenu setTitleBackgroundColor(int color) {
        return super.setTitleBackgroundColor(color);
    }

    @Override
    public WidgetMenu setTitleBackgroundColorRes(int color) {
        return super.setTitleBackgroundColorRes(color);
    }

    //
    //  Item
    //

    private Item buildItem;
    private boolean skipThisItem = false;
    private boolean skipGroup = false;

    private void finishItemBuilding() {
        if (buildItem != null) {
            Item i = buildItem;
            buildItem = null;
            if (!skipThisItem && !skipGroup) add(i);
            skipThisItem = false;
        }
    }

    public WidgetMenu add(@StringRes int text) {
        return add(ToolsResources.getString(text), null);
    }

    public WidgetMenu add(String text) {
        return add(text, null);
    }

    public WidgetMenu add(@StringRes int text, Callback1<WidgetMenu> onClick) {
        return add(ToolsResources.getString(text), onClick);
    }

    public WidgetMenu add(String text, Callback1<WidgetMenu> onClick) {
        finishItemBuilding();
        buildItem = new Item();
        buildItem.text = text;
        buildItem.onClick = onClick;
        return this;
    }

    public WidgetMenu text(@StringRes int text) {
        return text(ToolsResources.getString(text));
    }

    public WidgetMenu text(String text) {
        buildItem.text = text;
        return this;
    }

    public WidgetMenu icon(int icon) {
        buildItem.icon = icon;
        return this;
    }

    public WidgetMenu backgroundRes(@ColorRes int color) {
        return background(ToolsResources.getColor(color));
    }

    public WidgetMenu backgroundRes(@ColorRes int color, Provider<Boolean> condition) {
        if (condition.provide()) return background(ToolsResources.getColor(color));
        else return this;
    }

    public WidgetMenu background(@ColorInt int color) {
        buildItem.bg = color;
        return this;
    }

    public WidgetMenu preferred(boolean b) {
        buildItem.preferred = b;
        return this;
    }

    public WidgetMenu onClick(Callback1<WidgetMenu> onClick) {
        buildItem.onClick = onClick;
        return this;
    }

    public WidgetMenu condition(boolean b) {
        skipThisItem = !b;
        return this;
    }

    public WidgetMenu groupCondition(boolean b) {
        finishItemBuilding();
        skipGroup = !b;
        return this;
    }

    public WidgetMenu reverseGroupCondition() {
        finishItemBuilding();
        skipGroup = !skipGroup;
        return this;
    }

    public WidgetMenu clearGroupCondition() {
        finishItemBuilding();
        skipGroup = false;
        return this;
    }

    //
    //  Item
    //

    private class Item {

        private CardMenu card;
        private Callback1<WidgetMenu> onClick;
        private String text;
        private int icon;
        private int bg;
        private boolean preferred;

    }


}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/widgets/WidgetMenu.kt
package com.sup.dev.android.views.widgets

import android.support.annotation.ColorInt
import android.support.annotation.ColorRes
import android.support.annotation.StringRes
import android.support.v7.widget.LinearLayoutManager
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.views.support.adapters.recycler_view.RecyclerCardAdapter
import com.sup.dev.android.views.cards.CardDivider
import com.sup.dev.android.views.cards.CardDividerTitle
import com.sup.dev.android.views.cards.CardMenu

open class WidgetMenu : WidgetRecycler() {

    val myAdapter: RecyclerCardAdapter = RecyclerCardAdapter()
    private var onGlobalSelected: (WidgetMenu, String?) -> Unit = { w, s -> }

    private var prefCount = 0
    private var autoHide = true

    init {
        vRecycler.layoutManager = LinearLayoutManager(view.context)
        setAdapter<WidgetRecycler>(myAdapter)
    }

    override fun onShow() {
        vRecycler.isVerticalScrollBarEnabled = true
        super.onShow()
        finishItemBuilding()
    }

    fun setAutoHide(b: Boolean): WidgetMenu {
        autoHide = b
        return this
    }

    override fun setTitle(title: Int): WidgetMenu {
        super.setTitle(title)
        return this
    }

    override fun setTitle(title: String?): WidgetMenu {
        super.setTitle(title)
        return this
    }

    override fun setTitleBackgroundColor(color: Int): WidgetMenu {
        super.setTitleBackgroundColor(color)
        return this
    }

    override fun setTitleBackgroundColorRes(color: Int): WidgetMenu {
        super.setTitleBackgroundColorRes(color)
        return this
    }

    //
    //  Item
    //

    private var buildItem: Item? = null
    private var skipThisItem = false
    private var skipGroup = false


    fun clear() {
        finishItemBuilding()
        myAdapter.clear()
        prefCount = 0
    }

    private fun add(item: Item) {

        item.card = CardMenu()
        item.card?.text = item.text
        item.card?.setIcon(item.icon)
        item.card?.setBackground(item.bg)
        item.card?.setOnClick { v, x, y ->
            item.onClick.invoke(this, item.card!!)
            onGlobalSelected.invoke(this, item.text)
            if (autoHide) hide()
        }


        if (item.preferred) {
            if (prefCount == 0) myAdapter.add(0, CardDivider())
            myAdapter.add(prefCount, item.card!!)
            prefCount++
        } else {
            myAdapter.add(item.card!!)
        }

    }

    fun group(@StringRes title: Int): WidgetMenu {
        return group(ToolsResources.getString(title))
    }

    fun group(@StringRes title: Int, divider: Boolean): WidgetMenu {
        return group(ToolsResources.getString(title), divider)
    }

    @JvmOverloads
    fun group(title: String?, divider: Boolean = true): WidgetMenu {
        finishItemBuilding()
        myAdapter.add(CardDividerTitle().setText(title).setDividerBottom(divider))
        return this
    }

    fun setOnGlobalSelected(onGlobalSelected: (WidgetMenu, String?) -> Unit) {
        this.onGlobalSelected = onGlobalSelected
    }

    private fun finishItemBuilding() {
        if (buildItem != null) {
            val i = buildItem
            buildItem = null
            if (!skipThisItem && !skipGroup) add(i!!)
            skipThisItem = false
        }
    }

    fun add(@StringRes text: Int): WidgetMenu {
        return add(ToolsResources.getString(text))
    }

    fun add(@StringRes text: Int, onClick: (WidgetMenu, CardMenu) -> Unit = { w, c -> }): WidgetMenu {
        return add(ToolsResources.getString(text), onClick)
    }

    fun add(text: String, onClick: (WidgetMenu, CardMenu) -> Unit = { w, c -> }): WidgetMenu {
        finishItemBuilding()
        buildItem = Item()
        buildItem!!.text = text
        buildItem!!.onClick = onClick
        return this
    }

    fun text(@StringRes text: Int): WidgetMenu {
        return text(ToolsResources.getString(text))
    }

    fun text(text: String): WidgetMenu {
        buildItem!!.text = text
        return this
    }

    fun icon(icon: Int): WidgetMenu {
        buildItem!!.icon = icon
        return this
    }

    fun backgroundRes(@ColorRes color: Int): WidgetMenu {
        return background(ToolsResources.getColor(color))
    }

    fun backgroundRes(@ColorRes color: Int, condition: () -> Boolean): WidgetMenu {
        return if (condition.invoke())
            background(ToolsResources.getColor(color))
        else
            this
    }

    fun background(@ColorInt color: Int): WidgetMenu {
        buildItem!!.bg = color
        return this
    }

    fun preferred(b: Boolean): WidgetMenu {
        buildItem!!.preferred = b
        return this
    }

    fun onClick(onClick: (WidgetMenu, CardMenu) -> Unit): WidgetMenu {
        buildItem!!.onClick = onClick
        return this
    }

    fun condition(b: Boolean): WidgetMenu {
        skipThisItem = !b
        return this
    }

    fun groupCondition(b: Boolean): WidgetMenu {
        finishItemBuilding()
        skipGroup = !b
        return this
    }

    fun reverseGroupCondition(): WidgetMenu {
        finishItemBuilding()
        skipGroup = !skipGroup
        return this
    }

    fun clearGroupCondition(): WidgetMenu {
        finishItemBuilding()
        skipGroup = false
        return this
    }

    //
    //  Item
    //

    private inner class Item {

        var card: CardMenu? = null
        var onClick: (WidgetMenu, CardMenu) -> Unit = { w, c -> }
        var text: String = ""
        var icon: Int = 0
        var bg: Int = 0
        var preferred: Boolean = false

    }


}
